zenoh-kotlin/src/commonMain/kotlin/io/zenoh/prelude/QoS.kt

```diff
@@ -0,0 +1,60 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+package io.zenoh.prelude
+
+import io.zenoh.prelude.CongestionControl
+import io.zenoh.prelude.Priority
+
+/**
+ * Quality of service settings used to send zenoh message.
+ */
+class QoS internal constructor(internal val qos: Byte) {
+
+    /**
+     * Returns priority of the message.
+     */
+    fun priority(): Priority {
+        return Priority.fromInt(getPriorityViaJNI(qos))
+    }
+
+    /**
+     * Returns congestion control setting of the message.
+     */
+    fun congestionControl(): CongestionControl {
+        return CongestionControl.fromInt(getCongestionControlViaJNI(qos))
+    }
+
+    /**
+     * Returns express flag. If it is true, the message is not batched to reduce the latency.
+     */
+    fun express(): Boolean {
+        return getExpressViaJNI(qos);
+    }
+
+    companion object {
+        /**
+        * Returns default QoS settings.
+        */
+        fun default(): QoS {
+            return QoS(getDefaultViaJNI())
+        } 
+
+        private external fun getDefaultViaJNI(): Byte
+    }
+
+    private external fun getPriorityViaJNI(_qos: Byte): Int
+    private external fun getCongestionControlViaJNI(_qos: Byte): Int
+    private external fun getExpressViaJNI(_qos:Byte): Boolean
+}
```

Comment:
Let me suggest some changes here. So far all the communication through the JNI logic has been subjected to a delegation pattern, in which all the "ViaJNI" functions are contained within "JNI" objects on the 'jni' package in order to centralize that kind of logic there. So for the sake of sticking to that architectural design I'd do:


```suggestion
/**
 * Quality of service settings used to send zenoh message.
 */
class QoS internal constructor(private val jniQoS: JNIQoS) {

    internal constructor(qos: Byte): this(JNIQoS(qos))

    /**
     * Returns priority of the message.
     */
    fun priority(): Priority = jniQoS.getPriority()

    /**
     * Returns congestion control setting of the message.
     */
    fun congestionControl(): CongestionControl = jniQoS.getCongestionControl()

    /**
     * Returns express flag. If it is true, the message is not batched to reduce the latency.
     */
    fun express(): Boolean = jniQoS.getExpress() 

    companion object {
        /**
        * Returns default QoS settings.
        */
        fun default(): QoS {
            return QoS(JNIQoS())
        } 
    }
}
```

and where on the `JNIQoS` class contained within the jni package you'd have something like:

```kotlin
internal class JNIQoS internal constructor(private val qos: Byte) {

    internal constructor(): this(getDefaultQoSViaJNI())

    fun getExpress(): Boolean {
        return getExpressViaJNI(qos)
    }

    fun getCongestionControl(): CongestionControl {
        return CongestionControl.fromInt(getCongestionControlViaJNI(qos))
    }

    fun getPriority(): Priority {
        return Priority.fromInt(getPriorityViaJNI(qos))
    }

    private external fun getDefaultQoSViaJNI(): Byte
    private external fun getPriorityViaJNI(_qos: Byte): Int
    private external fun getCongestionControlViaJNI(_qos: Byte): Int
    private external fun getExpressViaJNI(_qos:Byte): Boolean
}
```

Reply:
Done